### PR TITLE
Remove no-op SDPA masks before backend dispatch

### LINKED ISSUE #177712: [dynamo] [torch.compile] sdpa become slower after torch._dynamo.run in dynamo benchmark when attention_mask is None
### 🐛 Describe the bug

`python benchmarks/dynamo/huggingface.py --performance --bfloat16 -dcuda -n3 --no-skip --dashboard --batch-size 2  --training --total-partitions 1 --partition-id 0 --only AlbertForQuestionAnswering --backend=eager --timeout=7200 --disable-cudagraphs  --export-profiler-trace` I run this on 4080s for `backend=eager` or `indcutor`.
In the exported trace, I can see it calls `aten::_flash_attention_forward` in eager run but calls `aten::_efficient_attention_forward` in the dynamo run. `attention_mask` changes from `None` to a tensor of all `True`

This will hurt performance, i.e. The total device time of sdpa bwd changes from 3.6ms to 6.4ms, fwd changes from 1.4ms to 2.0ms. Why we set attention_mask even it's None?

sdpa in trace marked as expected:
Image

sdpa in trace marked as actual:
Image

### Versions

torch                    2.12.0.dev20260317+cu128

cc @jerryzh168 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @Lucaskabela @jataylo @drisspg @liangel-02 @howardzhang-cv

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #184747

Dynamo can trace library masking helpers that materialize full-attention masks to avoid data-dependent checks, but a non-null mask prevents FlashAttention dispatch. Canonicalize structurally proven all-true boolean masks and zero additive masks to None before eager/Inductor backend execution.

Fixes #177712

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I added the `release notes: dynamo` label, marked the PR ready for review, and requested reviewers.

**Comment by jansel** [comment]
My agent says the Auto Request Review workflow had a transient `Bad credentials` failure after the PR was marked ready; I reran the failed job and it now passes.

**Comment by jansel** [comment]
My agent says I requested yushangdi as an additional reviewer because the PR has been waiting for review for over 48 hours; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
My agent says I requested karthickai as an additional reviewer because the PR has been waiting for review for more than 48 hours; drisspg remains requested as the attention expert reviewer.

**Comment by jansel** [comment]
My agent says I requested choijon5 as an additional reviewer because the PR has been waiting for review for more than 48 hours; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
My agent says I requested rupengliu-meta as an additional reviewer because the PR has been waiting for review for more than 48 hours; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
My agent says I verified the requested-reviewers list and requested BoyuanFeng as the additional reviewer; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
My agent says I requested laithsakka as an additional reviewer because the PR has been waiting for review for more than 48 hours; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
My agent says I requested zou3519 as an additional reviewer because the PR has been waiting for review for more than 48 hours and still has fewer than 8 requested reviewers; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
My agent says I requested frgossen as an additional reviewer because the PR has been waiting for review for more than 48 hours and still has fewer than 16 requested reviewers; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
My agent says I requested oulgen as an additional reviewer because the PR has been waiting for review for more than 48 hours and still has fewer than 16 requested reviewers; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
My agent says I requested mlazos as an additional reviewer because the PR has been waiting for review for more than 48 hours and still has fewer than 16 requested reviewers; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
My agent says I requested bobrenjc93 as an additional reviewer because the PR has been waiting for review for more than 48 hours and still has fewer than 16 requested reviewers; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
My agent says I requested williamwen42 as an additional reviewer because the PR has been waiting for review for more than 48 hours and still has fewer than 16 requested reviewers; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
My agent says I requested AmesingFlank as an additional reviewer because the PR has been waiting for review for more than 48 hours and still has fewer than 16 requested reviewers; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
My agent says I requested ethche as an additional reviewer because the PR has been waiting for review for more than 48 hours and still has fewer than 16 requested reviewers; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
My agent says I requested eellison as an additional reviewer because the PR has been waiting for review for more than 48 hours and still had fewer than 16 requested reviewers; drisspg remains requested as the SDPA expert reviewer.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and narrowed the optimization to the two exact Transformers mask producer graphs. The update adds metadata, purity, and side-effect guards plus regressions for causal calls, invalid dtype/device/layout cases, scalar casting and overflow, mutation/out= behavior, and both eager and Inductor FlashAttention dispatch. `python test/dynamo/test_sdpa.py` (24 tests) and `lintrunner -a` pass locally.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
